{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "metadata": {}
   },
   "outputs": [],
   "source": [
    "# Step 1: Prepare essential modules: FoldSeekAlignment, FoldSeekAlignmentParser\n",
    "\n",
    "\n",
    "class FoldSeekAlignment:\n",
    "    def __init__(self, line):\n",
    "        fields = line.strip().split('\\t')\n",
    "        \n",
    "        if len(fields) == 21:\n",
    "            # for most databases\n",
    "            self.qseqid = fields[0]\n",
    "            self.tseqid = fields[1]\n",
    "            self.pident = float(fields[2])\n",
    "            self.alnlen = int(fields[3])\n",
    "            self.mismatch = int(fields[4])\n",
    "            self.gapopen = int(fields[5])\n",
    "            self.qstart = int(fields[6])\n",
    "            self.qend = int(fields[7])\n",
    "            self.tstart = int(fields[8])\n",
    "            self.tend = int(fields[9])\n",
    "            self.evalue = float(fields[10])\n",
    "            self.bitscore = float(fields[11])\n",
    "            self.prob = int(fields[12])\n",
    "            self.qlen = int(fields[13])\n",
    "            self.tlen = int(fields[14])\n",
    "            self.qaln = fields[15]\n",
    "            self.taln = fields[16]\n",
    "            self.tca = [float(x) for x in fields[17].split(',')]\n",
    "            self.tseq = fields[18]\n",
    "            self.ttaxid = fields[19]\n",
    "            self.ttaxname = fields[20]\n",
    "        elif len(fields) == 19:\n",
    "            # for GMGC and MGnify\n",
    "            self.qseqid = fields[0]\n",
    "            self.tseqid = fields[1]\n",
    "            self.pident = float(fields[2])\n",
    "            self.alnlen = int(fields[3])\n",
    "            self.mismatch = int(fields[4])\n",
    "            self.gapopen = int(fields[5])\n",
    "            self.qstart = int(fields[6])\n",
    "            self.qend = int(fields[7])\n",
    "            self.tstart = int(fields[8])\n",
    "            self.tend = int(fields[9])\n",
    "            self.evalue = float(fields[10])\n",
    "            self.bitscore = float(fields[11])\n",
    "            self.prob = int(fields[12])\n",
    "            self.qlen = int(fields[13])\n",
    "            self.tlen = int(fields[14])\n",
    "            self.qaln = fields[15]\n",
    "            self.taln = fields[16]\n",
    "            self.tca = [float(x) for x in fields[17].split(',')]\n",
    "            self.tseq = fields[18]\n",
    "            self.ttaxid = None\n",
    "            self.ttaxname = None\n",
    "        else:\n",
    "            raise ValueError(\"Invalid FoldSeek .m8 line format\")\n",
    "\n",
    "    def __str__(self):\n",
    "        return f\"Query: {self.query_id}, target: {self.pident}, E-value: {self.evalue}, Bit Score: {self.bit_score}\"\n",
    "    \n",
    "    def print_alignment(self):\n",
    "        print(f\"Query: {self.qaln}\")\n",
    "        print(f\"Targt: {self.taln}\")\n",
    "\n",
    "\n",
    "\n",
    "class FoldSeekAlignmentParser:\n",
    "    def __init__(self, filename):\n",
    "        self.filename = filename\n",
    "\n",
    "    def parse(self):\n",
    "        with open(self.filename, 'r') as f:\n",
    "            alignments = [FoldSeekAlignment(line) for line in f.readlines()]\n",
    "        return alignments\n",
    "\n",
    "\n",
    "\n",
    "def check_reasonability(seq):\n",
    "    alphabet = set(seq)\n",
    "    if alphabet - set(\"ACDEFGHIKLMNPQRSTVWY\") != set():\n",
    "        return False\n",
    "    return True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "metadata": {}
   },
   "outputs": [],
   "source": [
    "# Step 2: input starting point, protected region\n",
    "folder_path = \"EPa23\"\n",
    "\n",
    "# set protected region\n",
    "protect_start = 9\n",
    "protect_end = 159"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "metadata": {}
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Database: afdb_proteome , found alignments: 9\n",
      "Database: afdb_swissprot , found alignments: 10\n",
      "Database: afdb50 , found alignments: 993\n",
      "Database: cath50 , found alignments: 14\n",
      "Database: gmgcl_id , found alignments: 66\n",
      "Database: mgnify_esm30 , found alignments: 998\n",
      "Database: pdb100 , found alignments: 74\n",
      "=====================================\n",
      "Total fraction of select alignments: 0.7980591497227357\n",
      "Total number of select alignments: 1727\n"
     ]
    }
   ],
   "source": [
    "# Step 3: Read and parse the alignment file\n",
    "folder_path = folder_path\n",
    "\n",
    "filename_list = [\n",
    "    folder_path + \"/alis_afdb-proteome.m8\",\n",
    "    folder_path + \"/alis_afdb-swissprot.m8\",\n",
    "    folder_path + \"/alis_afdb50.m8\",\n",
    "    folder_path + \"/alis_cath50.m8\",\n",
    "    folder_path + \"/alis_gmgcl_id.m8\",\n",
    "    folder_path + \"/alis_mgnify_esm30.m8\",\n",
    "    folder_path + \"/alis_pdb100.m8\"\n",
    "]\n",
    "\n",
    "alignments_collection = []\n",
    "alignments_dbname = [\n",
    "    \"afdb_proteome\",\n",
    "    \"afdb_swissprot\",\n",
    "    \"afdb50\",\n",
    "    \"cath50\",\n",
    "    \"gmgcl_id\",\n",
    "    \"mgnify_esm30\",\n",
    "    \"pdb100\"\n",
    "]\n",
    "\n",
    "# Parse the FoldSeek result files\n",
    "for filename in filename_list:\n",
    "    parser = FoldSeekAlignmentParser(filename)\n",
    "    alignments = parser.parse()\n",
    "    alignments_collection.append(alignments)\n",
    "\n",
    "total_alignments = 0\n",
    "# print results\n",
    "for i in range(len(alignments_dbname)):\n",
    "    print(\"Database:\",alignments_dbname[i],\", found alignments:\",len(alignments_collection[i]))\n",
    "    total_alignments += len(alignments_collection[i])\n",
    "\n",
    "\n",
    "print(\"=====================================\")\n",
    "\n",
    "\n",
    "# save to fasta\n",
    "f_out = open(f\"{folder_path}/{folder_path}_foldseek.fasta\",\"w\")\n",
    "count = 0\n",
    "\n",
    "for alignments_index in range(len(alignments_dbname)):\n",
    "    alignments = alignments_collection[alignments_index]\n",
    "    alignments_db = alignments_dbname[alignments_index]\n",
    "\n",
    "    for alignment in alignments:\n",
    "        # consider protect region\n",
    "        if alignment.qstart <= protect_start and alignment.qend >= protect_end:\n",
    "            f_out.write(\">\" + alignments_db + \" \" + alignment.tseqid.split(\" \")[0] + \"\\n\")\n",
    "            f_out.write(alignment.tseq + \"\\n\")\n",
    "            count += 1\n",
    "\n",
    "print(\"Total fraction of select alignments:\",count/total_alignments)\n",
    "print(\"Total number of select alignments:\",count)\n",
    "\n",
    "f_out.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "default",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
